File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        Map<String, Integer> participantMap =
            new HashMap<>();
        
        for (String person: participant) {
            if (participantMap.get(person) == null) {
                participantMap.put(person, 1);
            } else participantMap.merge(person, 1, Integer::sum);
        }
        
        for (String person: completion) {
            participantMap.merge(person, 1, (a, b) -> a - b);
            if (participantMap.get(person) == 0) {
                participantMap.remove(person);
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (String key: participantMap.keySet()) {
            sb.append(key);
        }
        return sb.toString();
    }
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.js
function solution(participant, completion) {
    const participantMap = new Map();
    
    for (const person of participant) {
        let count = participantMap.get(person) || 0;
        participantMap.set(person, count + 1);
    }
    
    for (const person of completion) {
        let count = participantMap.get(person);
        if (count == 1) participantMap.delete(person);
        else participantMap.set(person, count - 1);
    }
    
    let answer;
    for (const [key, value] of participantMap) {
        answer = key;
    }
    return answer;
}